src/Spago/Command/Fetch.purs

```diff
@@ -341,16 +372,47 @@ type TransitiveDepsResult =
       }
   }
 
-getTransitiveDeps :: forall a. Dependencies -> Spago (FetchEnv a) PackageMap
-getTransitiveDeps (Dependencies deps) = do
-  let depsRanges = map (fromMaybe Config.widestRange) deps
+getTransitiveDeps :: forall a. Config.WorkspacePackage -> Spago (FetchEnv a) PackageMap
+getTransitiveDeps workspacePackage = do
+  let depsRanges = map (fromMaybe Config.widestRange) (unwrap $ getWorkspacePackageDeps workspacePackage)
   { workspace } <- ask
-  case workspace.packageSet of
-    Registry extraPackages -> do
-      plan <- getTransitiveDepsFromRegistry depsRanges extraPackages
-      logDebug $ "Got a plan from the Solver: " <> printJson (Internal.Codec.packageMap Version.codec) plan
-      pure (map RegistryVersion plan)
-    PackageSet set -> getTransitiveDepsFromPackageSet set (Array.fromFoldable $ Map.keys depsRanges)
+  case workspace.packageSet.lockfile of
+    -- If we have a lockfile we can just use that - we don't even need build a plan, we can just filter the packages
+    -- marked as needed by the workspace package that we are processing, as we just dumped the plan in the lockfile.
```

Comment:
Computing the build plan when we have a lockfile is basically "filter by `needed_by`"